enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,463 @@
+
+### How Operators will opt-in
+
+Operators must opt-in to "OLM Supported Conditions" by adding an annotation to their owned CRD, where the key is the defined by OLM and the value is the `type` of the condition on the CR.
+
+#### Opt-in Example
+
+The "OLM Supported Condition" annotation key will adhere to the following format: `olm.Condition.<ConditionName>`. The value of the annotation will match the `Type` of the condition.  For example, the `NotInterruptible` condition's annotation key will be `olm.Condition.NotInterruptible`.
+
+While most "OLM Supported Condition" annotation keys will map to a single condition `type`, the `Important` condition will map to multiple condition types separated by commas.
+
+Let's consider an operator that installs a database. Assume that the operator already communicates that a database migration is in progress by setting the "Migrating" condition type in the status of the `foo` CustomResource. In this case, the operator should update the CRD to include the following annotation:
+
+```yaml
+"metadata": {
+  "annotations": {
+    "olm.Condition.NotInterruptible" : "Migrating",
+    "olm.Condition.Important" : "badConnectivity,unhealthyDatabase",
+  }
+}
+```
+
+By adding this annotation, OLM will search for all `foo` CustomResources and search for the `Migrating` Condition before upgrading the operator to the latest version.
+
+Now let's view the `foo` CR on cluster:
+
+```yaml
+apiVersion: foo.bar.com/v1
+kind: Foo
+metadata:
+  name: foo-database
+spec:
+  ...
+  ...
+  ...
+status:
+  conditions:
+  - type: Migrating # Condition 1
+    status: true
+    reason: "CriticalProcess" 
+    message: "Migrating old database schema to new format"
+  - type: unhealthyDatabase # Condition 2
+    status: true
+    reason: "LivenessProbeFailure"
+    message: "The foo-database failed its liveness probe."
+  - type: badConnectivity # Condition 2
+    status: true
+    reason: "404" 
+    message: "Unable to reach the database"
+```
+
+From the presence of the `Migrating` condition, OLM will understand that it should not upgrade the operator. The Operator also communicates that the `reason` the operator cannot be upgraded is due to a "critical process" and provides a human readable `message` for cluster admins.
+
+OLM will also highlight Conditions 1 and 2 within the new [Operator API](https://github.com/operator-framework/api/blob/master/crds/operators.coreos.com_operators.yaml) within the [condtions](https://github.com/operator-framework/api/blob/master/crds/operators.coreos.com_operators.yaml#L108) in the status.
+
+Note that by adding the `NotInterruptible` "OLM Supported Condition" to each CR, OLM will not upgrade the operator if any CR reports this condition as `true`.
+
+The complexity of communicating state of an operator can be reduced by having the operator introduce a new CRD which is used to communicate its conditions. This creates a single source of truth for the state of the operator.
```

Comment:
Is this saying "Each operator may just need to create a CRD that is OLM status" and that we are punting on defining that type?

If it is, then I would prefer we define it and can make it useful (see below comments).

Reply:
The idea is not to punt on defining the CRD, it is to avoid forcing operators to write to an OLM CRD. By doing so we:
- Make it easier for operators to integrate with OLM.
- Avoid expanding OLM's api surface.
- Avoid forcing operators to identify when they are ran with OLM.

Reply:
> Make it easier for operators to integrate with OLM.

I think you are helping an admin or someone package an operator for OLM to consume. I don't think that we are making it easier to integrate with OLM and easily communicate intent to OLM. Having to write a new API for my operator to do this, and then seeing 100's of other operators having to do the same is going to cause more confusion for each one of those operators. I would also think that it would cause confusion for admins, rather then having two known resources where status of the operator is written it is all over the place and sometimes in these API's defined by the library. I really think that an intermediate step of collection of the status to an API that can also be written to will be generically helpful for both cases. 

> Avoid expanding OLM's api surface.

This is true, we avoid OLM's api surface growing, but at the cost of asking 100's of operators to make a new API on their own. I think 1 new api for OLM is worth that trade off.

> Avoid forcing operators to identify when they are ran with OLM.

I am having hard time seeing how this does that. I think that if we went with the approach that I am asking for, it would be a simple "Does X API exist if not don't write status there". That seems like a reasonable thing to have operators do, especially when they want to tie to OLM. The folks who will write their own API want to be tightly integrated into OLM and I think we are making it harder for them to do that.

Reply:
+1 to what @shawn-hurley says here (and echoes what i said elsewhere in this review before seeing shawn's comments)

Reply:
These concerns should be resolved in the latest version of the PR.